server/rosetta/lib/internal/service/data.go

```diff
@@ -0,0 +1,158 @@
+package service
+
+import (
+	"context"
+
+	"github.com/coinbase/rosetta-sdk-go/types"
+	"github.com/cosmos/cosmos-sdk/server/rosetta/lib/errors"
+	crgtypes "github.com/cosmos/cosmos-sdk/server/rosetta/lib/types"
+)
+
+// AccountBalance retrieves the account balance of an address
+// rosetta requires us to fetch the block information too
+func (on OnlineNetwork) AccountBalance(ctx context.Context, request *types.AccountBalanceRequest) (*types.AccountBalanceResponse, *types.Error) {
+	var (
+		height int64
+		block  crgtypes.BlockResponse
+		err    error
+	)
+
+	switch {
+	case request.BlockIdentifier == nil:
+		block, err = on.client.BlockByHeight(ctx, nil)
+		if err != nil {
+			return nil, errors.ToRosetta(err)
+		}
+	case request.BlockIdentifier.Hash != nil:
+		block, err = on.client.BlockByHash(ctx, *request.BlockIdentifier.Hash)
+		if err != nil {
+			return nil, errors.ToRosetta(err)
+		}
+		height = block.Block.Index
+	case request.BlockIdentifier.Index != nil:
+		height = *request.BlockIdentifier.Index
+		block, err = on.client.BlockByHeight(ctx, &height)
+		if err != nil {
+			return nil, errors.ToRosetta(err)
+		}
+	}
+
+	accountCoins, err := on.client.Balances(ctx, request.AccountIdentifier.Address, &height)
+	if err != nil {
+		return nil, errors.ToRosetta(err)
+	}
+
+	return &types.AccountBalanceResponse{
+		BlockIdentifier: block.Block,
+		Balances:        accountCoins,
+		Metadata:        nil,
```

Comment:
I think adding token metadata in a follow up PR would be useful

Reply:
what do you have in mind specifically?

Reply:
Digging in, the [AccountBalanceResponse](https://www.rosetta-api.org/docs/models/AccountBalanceResponse.html) metadata is an optional field for the the account sequence number, ATM there is no place AFAIK where to store con metadata. 
I'll mark this as resolved for now